File: backend/students/utils/id_generator.py
# backend/students/utils/id_generator.py
class StudentIDGenerator:
    """Generate permanent student IDs in format: FS-YYYY-XXXXX"""
    
    def __init__(self):
        self.school_code = 'FS'
        
    def generate_student_id(self, existing_ids):
        """
        Generate a unique student ID
        Format: FS-2026-00001
        """
        import datetime
        current_year = datetime.datetime.now().year
        ethiopian_year = current_year - 8  # Approximate conversion
        
        # Find the highest sequence number
        max_seq = 0
        for existing_id in existing_ids:
            try:
                if existing_id and isinstance(existing_id, str):
                    parts = existing_id.split('-')
                    if len(parts) == 3:
                        seq = int(parts[2])
                        if seq > max_seq:
                            max_seq = seq
            except (ValueError, IndexError):
                continue
        
        # Generate next sequence
        next_seq = max_seq + 1
        padded_seq = str(next_seq).zfill(5)
        
        return f"{self.school_code}-{ethiopian_year}-{padded_seq}"

File: frontend/src/utils/idGenerator.js
// src/utils/idGenerator.js
class StudentIDGenerator {
  constructor() {
    this.schoolCode = 'FS';
    // Get current year in Ethiopian calendar (approx)
    const date = new Date();
    this.currentYear = date.getFullYear();
    this.ethiopianYear = this.currentYear - 8; // Approximate conversion
  }

  /**
   * Generate a unique student ID
   * Format: FS-YYYY-XXXXX (e.g., FS-2016-00001)
   */
  generateStudentID(existingIds = []) {
    const year = this.ethiopianYear;
    
    // Find the next available sequence number
    let maxSeq = 0;
    
    existingIds.forEach(id => {
      if (id && typeof id === 'string') {
        const parts = id.split('-');
        if (parts.length === 3) {
          const seq = parseInt(parts[2]);
          if (!isNaN(seq) && seq > maxSeq) {
            maxSeq = seq;
          }
        }
      }
    });
    
    const nextSeq = maxSeq + 1;
    const paddedSeq = nextSeq.toString().padStart(5, '0');
    return `${this.schoolCode}-${year}-${paddedSeq}`;
  }

  /**
   * Generate multiple IDs for bulk import
   */
  generateBulkIDs(count, existingIds = []) {
    const ids = [];
    let year = this.ethiopianYear;
    
    // Find max sequence
    let maxSeq = 0;
    existingIds.forEach(id => {
      if (id && typeof id === 'string') {
        const parts = id.split('-');
        if (parts.length === 3) {
          const seq = parseInt(parts[2]);
          if (!isNaN(seq) && seq > maxSeq) {
            maxSeq = seq;
          }
        }
      }
    });
    
    let sequence = maxSeq + 1;
    
    for (let i = 0; i < count; i++) {
      const paddedSeq = sequence.toString().padStart(5, '0');
      const id = `${this.schoolCode}-${year}-${paddedSeq}`;
      ids.push(id);
      sequence++;
    }
    
    return ids;
  }
}

// Create a single instance to export
const idGenerator = new StudentIDGenerator();
export default idGenerator;